From: IrfanMohammad <irfan@graphiant.com>
Date: Thu, 28 Sep 2023 17:39:57 +0000
Subject: - Optimize xpath filtering for oper get subscr

 1. When the requested xpath is "/*" or "/module-name:*", we can
    short-circuit the check if the subscr is relevant.

 2. sysrepo-monitoring module oper may not need to check
    `sr_modinfo_module_data_oper_required` as often as it does.
    Most of the time, it is required.

 3. When a single module is queried for sysrepo-monitoring, don't
    process all modules.

 4. Don't acquire locks for modules without subscriptions for
    sysrepo-monitoring

- API to check sr_module_has_change_sub efficiently

Currently, the only way to check if a module has a subscription is via
sysrepo-monitoring and that is extremely inefficient.

When applications poll for a subscription to be created, it is better to
have a simple api to first check whether the module has any
subscriptions.
---
 src/modinfo.c                 | 120 +++++++++++++++++++++++++++++++++++++-----
 src/sysrepo.c                 |  78 +++++++++++++++++++++++++++
 src/sysrepo.h                 |   4 ++
 tests/test_subscr_filtering.c |  36 +++++++++++++
 4 files changed, 224 insertions(+), 14 deletions(-)

diff --git a/src/modinfo.c b/src/modinfo.c
index 35e47b2..010a362 100644
--- a/src/modinfo.c
+++ b/src/modinfo.c
@@ -826,6 +826,22 @@ sr_xpath_oper_data_text_atoms_required(const char *request_atom, const char *sub
     return 1;
 }
 
+static int
+sr_xpath_is_entire_module(const char *xpath)
+{
+    int mod_len = 0;
+
+    if (xpath[0] != '/') {
+        return 0;
+    }
+
+    /* get module name offset from xpath */
+    sr_xpath_next_qname(xpath + 1, NULL, &mod_len, NULL, NULL);
+
+    /* Looking for /module-name:* */
+    return (strcmp(xpath + mod_len + 1, ":*") == 0);
+}
+
 /**
  * @brief Check whether operational data are required.
  *
@@ -846,8 +862,8 @@ sr_xpath_oper_data_required(const char *request_xpath, const char *sub_xpath, in
 
     *required = 1;
 
-    if (!request_xpath) {
-        /* we do not know, say it is required */
+    if (!request_xpath || !strcmp(request_xpath, "/*") || sr_xpath_is_entire_module(request_xpath)) {
+        /* we do not know or everything or entire module requested, say it is required */
         goto cleanup;
     }
 
@@ -1718,6 +1734,11 @@ sr_modinfo_module_srmon_module_subs(sr_conn_ctx_t *conn, sr_mod_t *shm_mod, stru
      */
 
     for (ds = 0; ds < SR_DS_COUNT; ++ds) {
+        if (!shm_mod->change_sub[ds].sub_count) {
+            /* no subscribers to this module and ds */
+            continue;
+        }
+
         /* CHANGE SUB READ LOCK */
         if ((err_info = sr_rwlock(&shm_mod->change_sub[ds].lock, SR_SHMEXT_SUB_LOCK_TIMEOUT, SR_LOCK_READ, conn->cid,
                 __func__, NULL, NULL))) {
@@ -1778,6 +1799,9 @@ change_sub_unlock:
     /*
      * oper get subscriptions
      */
+    if (!shm_mod->oper_get_sub_count) {
+        goto oper_poll_subs;
+    }
 
     /* OPER GET SUB READ LOCK */
     if ((err_info = sr_rwlock(&shm_mod->oper_get_lock, SR_SHMEXT_SUB_LOCK_TIMEOUT, SR_LOCK_READ, conn->cid,
@@ -1828,9 +1852,13 @@ operget_sub_unlock:
         return err_info;
     }
 
+oper_poll_subs:
     /*
      * oper poll subscriptions
      */
+    if (!shm_mod->oper_poll_sub_count) {
+        goto notif_subs;
+    }
 
     /* OPER POLL SUB READ LOCK */
     if ((err_info = sr_rwlock(&shm_mod->oper_poll_lock, SR_SHMEXT_SUB_LOCK_TIMEOUT, SR_LOCK_READ, conn->cid,
@@ -1875,6 +1903,11 @@ operpoll_sub_unlock:
         return err_info;
     }
 
+notif_subs:
+    if (!shm_mod->notif_sub_count) {
+        return err_info;
+    }
+
     /*
      * notification subscriptions
      */
@@ -1929,7 +1962,7 @@ notif_sub_unlock:
  * @return err_info, NULL on success.
  */
 static sr_error_info_t *
-sr_modinfo_module_srmon_module(sr_conn_ctx_t *conn, sr_mod_t *shm_mod, struct lyd_node *sr_state)
+sr_modinfo_module_srmon_module(sr_conn_ctx_t *conn, sr_mod_t *shm_mod, struct lyd_node *sr_state, int change_subs_only)
 {
     sr_error_info_t *err_info = NULL;
     struct lyd_node *sr_mod, *sr_ds_lock;
@@ -1944,6 +1977,10 @@ sr_modinfo_module_srmon_module(sr_conn_ctx_t *conn, sr_mod_t *shm_mod, struct ly
     SR_CHECK_LY_RET(lyd_new_list(sr_state, NULL, "module", 0, &sr_mod, conn->mod_shm.addr + shm_mod->name), ly_ctx,
             err_info);
 
+    if (change_subs_only) {
+        goto subscriptions;
+    }
+
     /* last-modified */
     for (ds = 0; ds < SR_DS_COUNT; ++ds) {
         if ((err_info = sr_modinfo_module_srmon_datastore(conn, shm_mod, ds, sr_mod))) {
@@ -2017,6 +2054,7 @@ ds_unlock:
         return err_info;
     }
 
+subscriptions:
     /* module subscriptions */
     if ((err_info = sr_modinfo_module_srmon_module_subs(conn, shm_mod, sr_mod))) {
         return err_info;
@@ -2174,6 +2212,36 @@ cleanup:
     return err_info;
 }
 
+static int
+sr_modinfo_load_srmon_single_module(const char *xpath, char *mod_name, int *change_sub_only)
+{
+    char q1 = 0, q2 = 0;
+    char buf[128] = "";
+    int result, len;
+    *change_sub_only = 0;
+
+    if (strchr(xpath, '|')) {
+        return 0;
+    }
+
+    if (3 == sscanf(xpath, "/sysrepo-monitoring:sysrepo-state/module[name=%c%64[^\"\']%c]",
+                &q1, mod_name, &q2)) {
+        result = (q1 == q2) && (q1 == '\"' || q1 == '\'');
+        if (!result) {
+            return 0;
+        }
+
+        len = snprintf(buf, sizeof(buf), "/sysrepo-monitoring:sysrepo-state/module[name=\"%s\"]/subscriptions/change-sub/xpath", mod_name);
+
+        if (!strncmp(buf, xpath, len)) {
+            *change_sub_only = 1;
+        }
+        return 1;
+    }
+
+    return 0;
+}
+
 /**
  * @brief Load module data of the sysrepo-monitoring module. They are actually generated.
  *
@@ -2193,7 +2261,11 @@ sr_modinfo_module_data_load_srmon(struct sr_mod_info_s *mod_info, struct sr_mod_
     const struct lys_module *ly_mod;
     sr_mod_shm_t *mod_shm;
     uint32_t i, j;
-    int req;
+    int req, check = 1;
+    char req_mod_name[65] = "";
+    int change_subs_only = 0;
+
+    #define needed ((!check) || req)
 
     mod_shm = SR_CONN_MOD_SHM(mod_info->conn);
     ly_mod = ly_ctx_get_module_implemented(mod_info->conn->ly_ctx, "sysrepo-monitoring");
@@ -2202,39 +2274,56 @@ sr_modinfo_module_data_load_srmon(struct sr_mod_info_s *mod_info, struct sr_mod_
     /* main container */
     SR_CHECK_LY_GOTO(lyd_new_inner(NULL, ly_mod, "sysrepo-state", 0, &mod_data), mod_info->conn->ly_ctx, err_info, cleanup);
 
+    /* determine if entire sysrepo-monitoring was requested */
+    check = (strcmp(mod->xpaths[0], "/*") && strcmp(mod->xpaths[0], "/sysrepo-monitoring:*"));
+
+    /* check if we only need a single module */
+    if (check && sr_modinfo_load_srmon_single_module(mod->xpaths[0], req_mod_name, &change_subs_only)) {
+        shm_mod = sr_shmmod_find_module(mod_shm, req_mod_name);
+        if (!shm_mod) {
+            sr_errinfo_new(&err_info, SR_ERR_NOT_FOUND, "Module %s not found", req_mod_name);
+            goto cleanup;
+        }
+        if ((err_info = sr_modinfo_module_srmon_module(mod_info->conn, shm_mod, mod_data, change_subs_only))) {
+            goto cleanup;
+        }
+        goto finished;
+    }
+
     /* modules */
-    if ((err_info = sr_modinfo_module_data_oper_required(mod, &req, "/sysrepo-monitoring:sysrepo-state/module"))) {
+    if (check && (err_info = sr_modinfo_module_data_oper_required(mod, &req, "/sysrepo-monitoring:sysrepo-state/module"))) {
         goto cleanup;
     }
-    if (req) {
+
+    if (needed) {
         for (i = 0; i < mod_shm->mod_count; ++i) {
             shm_mod = SR_SHM_MOD_IDX(mod_shm, i);
-            if ((err_info = sr_modinfo_module_data_oper_required(mod, &req,
+            if (check && (err_info = sr_modinfo_module_data_oper_required(mod, &req,
                     "/sysrepo-monitoring:sysrepo-state/module[name='%s']", mod_info->conn->mod_shm.addr + shm_mod->name))) {
                 goto cleanup;
             }
 
-            if (req && (err_info = sr_modinfo_module_srmon_module(mod_info->conn, shm_mod, mod_data))) {
+            if (needed && (err_info = sr_modinfo_module_srmon_module(mod_info->conn, shm_mod, mod_data, 0))) {
                 goto cleanup;
             }
         }
     }
 
     /* RPCs */
-    if ((err_info = sr_modinfo_module_data_oper_required(mod, &req, "/sysrepo-monitoring:sysrepo-state/rpc"))) {
+    if (check && (err_info = sr_modinfo_module_data_oper_required(mod, &req, "/sysrepo-monitoring:sysrepo-state/rpc"))) {
         goto cleanup;
     }
-    if (req) {
+    if (needed) {
         for (i = 0; i < mod_shm->mod_count; ++i) {
             shm_mod = SR_SHM_MOD_IDX(mod_shm, i);
             shm_rpc = (sr_rpc_t *)(mod_info->conn->mod_shm.addr + shm_mod->rpcs);
             for (j = 0; j < shm_mod->rpc_count; ++j) {
-                if ((err_info = sr_modinfo_module_data_oper_required(mod, &req,
+                if (check && (err_info = sr_modinfo_module_data_oper_required(mod, &req,
                         "/sysrepo-monitoring:sysrepo-state/rpc[path='%s']", mod_info->conn->mod_shm.addr + shm_rpc[j].path))) {
                     goto cleanup;
                 }
 
-                if (req && (err_info = sr_modinfo_module_srmon_rpc(mod_info->conn, &shm_rpc[j], mod_data))) {
+                if (needed && (err_info = sr_modinfo_module_srmon_rpc(mod_info->conn, &shm_rpc[j], mod_data))) {
                     goto cleanup;
                 }
             }
@@ -2242,13 +2331,14 @@ sr_modinfo_module_data_load_srmon(struct sr_mod_info_s *mod_info, struct sr_mod_
     }
 
     /* connections */
-    if ((err_info = sr_modinfo_module_data_oper_required(mod, &req, "/sysrepo-monitoring:sysrepo-state/connection"))) {
+    if (check && (err_info = sr_modinfo_module_data_oper_required(mod, &req, "/sysrepo-monitoring:sysrepo-state/connection"))) {
         goto cleanup;
     }
-    if (req && (err_info = sr_modinfo_module_srmon_connections(mod_data))) {
+    if (needed && (err_info = sr_modinfo_module_srmon_connections(mod_data))) {
         goto cleanup;
     }
 
+finished:
     /* connect to the rest of data */
     if (lyd_merge_siblings(&mod_info->data, mod_data, LYD_MERGE_DESTRUCT)) {
         sr_errinfo_new_ly(&err_info, mod_info->conn->ly_ctx, NULL);
@@ -2256,6 +2346,8 @@ sr_modinfo_module_data_load_srmon(struct sr_mod_info_s *mod_info, struct sr_mod_
     }
     mod_data = NULL;
 
+#undef needed
+
 cleanup:
     lyd_free_siblings(mod_data);
     return err_info;
diff --git a/src/sysrepo.c b/src/sysrepo.c
index e63c1b6..c72affc 100644
--- a/src/sysrepo.c
+++ b/src/sysrepo.c
@@ -7407,3 +7407,81 @@ cleanup:
     }
 }
 
+API int
+sr_module_has_change_sub(sr_conn_ctx_t *conn, const char *module_name,
+        const sr_datastore_t ds, char ***xpaths, uint32_t *sub_count)
+{
+    char *xpath;
+    size_t i = 0;
+    sr_error_info_t *err_info = NULL;
+    sr_mod_change_sub_t *change_subs;
+
+    SR_CHECK_ARG_APIRET(!xpaths || !sub_count || !conn || !module_name ||
+            (ds > SR_DS_FACTORY_DEFAULT), NULL, err_info);
+
+    *xpaths = NULL;
+
+    /* CONTEXT LOCK */
+    if ((err_info = sr_lycc_lock(conn, SR_LOCK_READ, 0, __func__))) {
+        return sr_api_ret(NULL, err_info);
+    }
+
+    sr_mod_t *shm_mod = sr_shmmod_find_module(SR_CONN_MOD_SHM(conn), module_name);
+    if (!shm_mod) {
+        sr_errinfo_new(&err_info, SR_ERR_NOT_FOUND, "Module \"%s\" was not found in sysrepo.", module_name);
+        goto cleanup;
+    }
+
+    *sub_count = shm_mod->change_sub[ds].sub_count;
+    if (!*sub_count) {
+        goto cleanup;
+    }
+
+    /* CHANGE SUB READ LOCK */
+    if ((err_info = sr_rwlock(&shm_mod->change_sub[ds].lock, SR_SHMEXT_SUB_LOCK_TIMEOUT, SR_LOCK_READ, conn->cid,
+            __func__, NULL, NULL))) {
+        goto change_sub_unlock;
+    }
+
+    /* EXT READ LOCK */
+    if ((err_info = sr_shmext_conn_remap_lock(conn, SR_LOCK_READ, 0, __func__))) {
+        goto change_ext_sub_unlock;
+    }
+
+    change_subs = (sr_mod_change_sub_t *)(conn->ext_shm.addr + shm_mod->change_sub[ds].subs);
+
+    *xpaths = malloc(*sub_count * sizeof (*xpaths));
+
+    for (i = 0; i < *sub_count; i++) {
+        errno = 0;
+        xpath = conn->ext_shm.addr + change_subs[i].xpath;
+        (*xpaths)[i] = xpath ? strdup(xpath) : NULL;
+
+        if (errno) {
+            SR_ERRINFO_SYSERRNO(&err_info, "strdup");
+            sr_errinfo_new(&err_info, SR_ERR_INTERNAL, "Out of memory");
+            goto cleanup_mem;
+        }
+    }
+    goto change_ext_sub_unlock;
+
+cleanup_mem:
+    while (i) {
+        free((*xpaths)[i]);
+        i--;
+    }
+    free(*xpaths);
+    *xpaths = NULL;
+
+change_ext_sub_unlock:
+        /* EXT READ UNLOCK */
+        sr_shmext_conn_remap_unlock(conn, SR_LOCK_READ, 0, __func__);
+
+change_sub_unlock:
+        /* CHANGE SUB READ UNLOCK */
+        sr_rwunlock(&shm_mod->change_sub[ds].lock, SR_SHMEXT_SUB_LOCK_TIMEOUT, SR_LOCK_READ, conn->cid, __func__);
+
+cleanup:
+    sr_lycc_unlock(conn, SR_LOCK_READ, 0, __func__);
+    return sr_api_ret(NULL, err_info);
+}
diff --git a/src/sysrepo.h b/src/sysrepo.h
index ec54bd8..09f68e3 100644
--- a/src/sysrepo.h
+++ b/src/sysrepo.h
@@ -1895,6 +1895,10 @@ void sr_session_set_nc_id(sr_session_ctx_t *session, uint64_t nc_sid);
  * */
 int sr_module_rwlock_debug(sr_conn_ctx_t *conn, const char *module_name, uint32_t mask);
 
+int
+sr_module_has_change_sub(sr_conn_ctx_t *conn, const char *module_name,
+        const sr_datastore_t datastore, char ***xpaths, uint32_t *sub_count);
+
 #ifdef __cplusplus
 }
 #endif
diff --git a/tests/test_subscr_filtering.c b/tests/test_subscr_filtering.c
index e4ef857..5dc2312 100644
--- a/tests/test_subscr_filtering.c
+++ b/tests/test_subscr_filtering.c
@@ -165,6 +165,15 @@ module_error_cb(sr_session_ctx_t *session, uint32_t sub_id, const char *module_n
     TLOG_INF("%s called with %s %s", __func__, module_name, xpath);
     return SR_ERR_OPERATION_FAILED;
 }
+
+static void free_xpaths_arr(char **xpaths, size_t count)
+{
+    while (count--) {
+        free(xpaths[count]);
+    }
+    free(xpaths);
+}
+
 /**
  * Check that changes which have all xpaths filtered are not called back
  * tests ENG-5259
@@ -178,6 +187,8 @@ test_subscr_none(void **arg)
     int i = 0, ret;
     cb_stats stats[4];
     int expected_counts[] = {0, 0, 0, 1};
+    uint32_t count;
+    char **xpaths;
 
     sr_session_start(st->conn, SR_DS_RUNNING, &st->sess);
     memset(subscr, 0, sizeof(subscr));
@@ -232,6 +243,31 @@ test_subscr_none(void **arg)
         assert_int_equal(expected_counts[i], stats[i].count);
     }
 
+    ret = sr_module_has_change_sub(st->conn, "test", SR_DS_RUNNING, &xpaths, &count);
+    assert_int_equal(ret, SR_ERR_OK);
+    assert_int_equal(count, 3);
+    free_xpaths_arr(xpaths, count);
+
+    ret = sr_module_has_change_sub(st->conn, "test", SR_DS_OPERATIONAL, &xpaths, &count);
+    assert_int_equal(ret, SR_ERR_OK);
+    assert_int_equal(count, 0);
+    free_xpaths_arr(xpaths, count);
+
+    ret = sr_module_has_change_sub(st->conn, "test-leafref", SR_DS_RUNNING, &xpaths, &count);
+    assert_int_equal(ret, SR_ERR_OK);
+    assert_int_equal(count, 1);
+    free_xpaths_arr(xpaths, count);
+
+    ret = sr_module_has_change_sub(st->conn, "ietf-interfaces", SR_DS_RUNNING, &xpaths, &count);
+    assert_int_equal(ret, SR_ERR_OK);
+    assert_int_equal(count, 0);
+    free_xpaths_arr(xpaths, count);
+
+    ret = sr_module_has_change_sub(st->conn, "blargh", SR_DS_RUNNING, &xpaths, &count);
+    assert_int_equal(ret, SR_ERR_NOT_FOUND);
+    assert_int_equal(count, 0);
+    free_xpaths_arr(xpaths, count);
+
     for (i = 0; i < 4; i++) {
         sr_unsubscribe(subscr[i]);
     }
